Source/typescript/backend/web/Context.ts

```diff
@@ -1,8 +1,50 @@
 // Copyright (c) Dolittle. All rights reserved.
 // Licensed under the MIT license. See LICENSE file in the project root for full license information.
 
-export type Context = {
-    userId: string;
-    tenantId: string;
-    cookies: string;
+import cls from 'cls-hooked';
+import { Request, Response, NextFunction } from 'express';
+import { TenantId } from '@dolittle/sdk.execution';
+
+const ContextKey = 'Context';
+const ns = cls.createNamespace('0e852ce9-ef09-45d2-b64b-4868de226563');
```

Comment:
The key can be anything - as long as its unique.

What it does is give us a key/value store per async context - making it safe to store values and be sure you’re using it in the right context when you’re in the same call context. Similar to AsyncLocal in C#/.net.

Our use case is to be able for each web request coming in to set the correct context based on TenantId, UserId coming in and such and make this accessible in a good way throughout different parts of the request.

Its leveraging a project called `cls-hooked` for this - which makes it a bit nicer to do. You can read more on it here: https://medium.com/trabe/continuation-local-storage-for-easy-context-passing-in-node-js-2461c2120284

Reply:
That looks like a nice and unobtrusive way of adding this context information. I read the blog post and it links to some performance issues that may be worth considering, as this is adding a significant performance hit according to that Github thread. https://github.com/nodejs/benchmarking/issues/181

>You’d also want to stay away from async_hooks if you’re writing a library since you could impact performance for your users just for your convenience.

There doesn't seem to be any improvement or resolution though. Perhaps another approach may be better for this, or we could see how it actually impacts applications? Perhaps creating a follow-up task for this?

Reply:
Good to keep an eye on the performance. I've added an issue (#121) as you suggested.

Great catch! The benchmarks they had there was on version 9 of Node. I think for most of our current scenarios we are ok. 

https://github.com/bmeurer/async-hooks-performance-impact

It talks about # of requests a second in the thousands. 

<img width="312" alt="image" src="https://user-images.githubusercontent.com/134365/109283834-8116b380-781f-11eb-8f4b-f125229312d5.png">

![](https://raw.githubusercontent.com/bmeurer/async-hooks-performance-impact/master/results-hapi-koa-node-9.4.0.png)